### PR TITLE
API injections for socket groups

### PR DESCRIPTION
**Pull request by ethouris**
From #1090, point 2 - API changes.

API docs are in PR #1123

### CODE REVIEW

**Comment by maxsharabayko** [comment]
@ethouris What about updates to the API.md and the API-functions.md?
Are they in some separate PR? Please share its numbe if so.

**Comment by maxsharabayko** [comment]
### 1. Listener/caller workflow unification

On the caller side:
```
sockgroup_id = srt_create_group(SRT_GTYPE_BROADCAST);
srt_connect_group(sockgroup_id, address, ...);
```

On the listener side:
```
int sock = srt_create_socket();
srt_setsockflag(sock, SRTO_GROUPCONNECT, &yes, sizeof yes);
srt_listen(...)
```

Why can't the  `srt_create_group(SRT_GTYPE_REDUNDANCY)` be used instead of `srt_create_socket()` and `srt_setsockflag(...)` on the listener side?


### 2. Listener bonding several ports

With the current implementation and the API there is no way to have a listener bonded to different ports?

For example, accepting a bonded connection on 192.168.0.10:4200 and 10.129.10.121:4300. And limiting the nuber of connections on each interface.

**Comment by ethouris** [comment]
1. The problem is that a listener socket and communication socket are two completely different entities, mistakenly bound under one name (since BSD socket API, which UDT simply followed) - but they otherwise have nothing to do with one another. There are completely different sets of operations done on them, and even some common operations do completely different things, like:

* srt_bind: for listener sets up the connection point, for communication socket sets up the source

* socket options: those for listener socket do not mean anything for a communication socket. Those for communication socket are simply carried over by the listener socket into the accepted socket. A small number of socket options that are common (like `SRTO_RCVSYN`) needs usually to be altered once the accepted socket is returned

In a decent API, listener and communication socket would be two completely different entities, implemented using different data types.

In the group API, the "group" entity is a replacement for communication socket and nothing else. Same as you don't create a socket that the listener would use to handle the incoming connection, but the connecting process creates it by itself, same it is with the group.

First of all, the rules that you need a connection point - in the form of listener socket - are still in force. The listening socket is capable normally of returning multiple connections with multiple requests and there's a socket created for each of them.

By allowing for group connections, you simply extend that functionality so that this listener can return **both single sockets and groups**, as well as it will create a new group when the first group connection is attached. Next connections within the same group will be handled in background. But next connections with a **different** group will also create a new group, just like it creates a new socket with the new incoming connection.

2. Listener can only listen on one port and all devices, or one port on a selected device. This functionality could be extended to bind multiple binding-points in one listener, but this is rather a bigger project that involves using multiple UDP sockets by one listener, and then possibly also using multiple SndQ and RcvQ's, of which only the right one would be shared with an accepted socket etc., and I doubt that it's worth it because this can be easily manipulated by playing around with virtual devices and routing rules.

File: srtcore/api.cpp [hunk 1]
```diff
@@ -684,9 +800,6 @@ int CUDTUnited::newConnection(const SRTSOCKET listen, const sockaddr_any& peer,
       return -1;
    }
 
-   // wake up a waiting accept() call
-   CSync::lock_signal(ls->m_AcceptCond, ls->m_AcceptLock);
-
    return 1;
```
Comment (maxsharabayko) [inline srtcore/api.cpp]: Why removed?

File: srtcore/api.cpp [hunk 2]
```diff
@@ -1034,6 +1519,16 @@ int CUDTUnited::connectIn(CUDTSocket* s, const sockaddr_any& target_addr, int32_
...
+    {
+        CUDTGroup* g = locateGroup(u);
+        if (!g)
+            throw CUDTException(MJ_NOTSUP, MN_SIDINVAL, 0);
+
+        g->close();
+        deleteGroup(g);
+        return 0;
+    }
    CUDTSocket* s = locateSocket(u);
```
Comment (maxsharabayko) [inline srtcore/api.cpp]: Alignment looks wrong

File: srtcore/api.cpp [hunk 3]
```diff
@@ -1087,6 +1582,7 @@ int CUDTUnited::close(CUDTSocket* s)
 
        // since "s" is located before m_GlobControlLock, locate it again in case
        // it became invalid
+       // XXX Seriously? Why not do all things necessary on s prior to deletion?
```
Comment (maxsharabayko) [inline srtcore/api.cpp]: Better to replace with guidance on improvement.

File: srtcore/core.cpp [hunk 4]
```diff
@@ -858,6 +860,13 @@ void CUDT::setOpt(SRT_SOCKOPT optName, const void* optval, int optlen)
...
         break;
 
+
+   case SRTO_GROUPCONNECT:
+        if (m_bConnected)
+            throw CUDTException(MJ_NOTSUP, MN_ISCONNECTED, 0);
+        m_bOPT_GroupConnect = bool_int_value(optval, optlen);
+        break;
+
     case SRTO_KMREFRESHRATE:
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: formatting

File: srtcore/api.cpp (hunk 1 above)
Comment (ethouris) [reply srtcore/api.cpp]: It's moved, not removed. See line 756.

File: srtcore/api.cpp (hunk 2 above)
Comment (ethouris) [reply srtcore/api.cpp]: Reformatted as a whole

File: srtcore/api.cpp (hunk 3 above)
Comment (ethouris) [reply srtcore/api.cpp]: Improved.

**Pull request merged**